src/transformers/modeling_longformer.py

```diff
@@ -0,0 +1,667 @@
+        self.head_dim = int(config.hidden_size / config.num_attention_heads)
+        self.embed_dim = config.hidden_size
+
+        self.query = nn.Linear(config.hidden_size, self.embed_dim)
+        self.key = nn.Linear(config.hidden_size, self.embed_dim)
+        self.value = nn.Linear(config.hidden_size, self.embed_dim)
+
+        # separate projection layers for tokens with global attention
+        self.query_global = nn.Linear(config.hidden_size, self.embed_dim)
+        self.key_global = nn.Linear(config.hidden_size, self.embed_dim)
+        self.value_global = nn.Linear(config.hidden_size, self.embed_dim)
+
+        self.dropout = config.attention_probs_dropout_prob
+
+        self.layer_id = layer_id
+        attention_window = config.attention_window[self.layer_id]
+        assert attention_window % 2 == 0, f'`attention_window` for layer {self.layer_id} has to be an even value. Given {attention_window}'
+        assert attention_window > 0, f'`attention_window` for layer {self.layer_id} has to be positive. Given {attention_window}'
+
+        self.one_sided_attention_window_size = attention_window // 2
+
+    def _skew(self, x, direction, padding_value):
+        '''Convert diagonals into columns (or columns into diagonals depending on `direction`'''
+        x_padded = F.pad(x, direction, value=padding_value)
+        x_padded = x_padded.view(*x_padded.size()[:-2], x_padded.size(-1), x_padded.size(-2))
+        return x_padded
+
+    def _skew2(self, x, padding_value):
+        '''shift every row 1 step to right converting columns into diagonals'''
+        # X = B x C x M x L
+        B, C, M, L = x.size()
+        x = F.pad(x, (0, M + 1), value=padding_value)  # B x C x M x (L+M+1)
+        x = x.view(B, C, -1)  # B x C x ML+MM+M
+        x = x[:, :, :-M]  # B x C x ML+MM
+        x = x.view(B, C, M, M + L)  # B x C, M x L+M
+        x = x[:, :, :, :-1]
+        return x
+
+    def _chunk(self, x, w):
+        '''convert into overlapping chunkings. Chunk size = 2w, overlap size = w'''
+
+        # non-overlapping chunks of size = 2w
+        x = x.view(x.size(0), x.size(1) // (w * 2), w * 2, x.size(2))
+
+        # use `as_strided` to make the chunks overlap with an overlap size = w
+        chunk_size = list(x.size())
+        chunk_size[1] = chunk_size[1] * 2 - 1
+
+        chunk_stride = list(x.stride())
+        chunk_stride[1] = chunk_stride[1] // 2
+        return x.as_strided(size=chunk_size, stride=chunk_stride)
+
+    @lru_cache()
+    def _get_invalid_locations_mask(self, w, device):
+        affected_seq_len = w
+        diagonals_list = []
+        for j in range(-w, 1, 1):
+            diagonal_mask = torch.zeros(affected_seq_len, device='cpu', dtype=torch.uint8)
+            diagonal_mask[:-j] = 1
+            diagonals_list.append(diagonal_mask)
+        mask = torch.stack(diagonals_list, dim=-1)
+        mask = mask[None, :, None, :]
+        ending_mask = mask.flip(dims=(1, 3)).bool().to(device)
```

Comment:
@sshleifer @thomwolf - I think `bool()` break backward compatibility, no? I think we still use `torch.uint8` instead at the moment. So i think we should change `.bool()` to `.to(torch.uint8)` here